unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -206,6 +207,198 @@ public UTF8String toLowerCase() {
     return fromString(toString().toLowerCase());
   }
 
+  /**
+   * Copy the bytes from the current UTF8String, and make a new UTF8String.
+   * @param start the start position of the current UTF8String in bytes.
+   * @param end the end position of the current UTF8String in bytes.
+   * @return a new UTF8String in the position of [start, end] of current UTF8String bytes.
+   */
+  private UTF8String copyUTF8String(byte[] bytes, int start, int end) {
+    int len = end - start + 1;
+    byte[] newBytes = new byte[len];
+    System.arraycopy(bytes, start, newBytes, 0, len);
+    return UTF8String.fromBytes(newBytes);
+  }
+
+  public UTF8String trim() {
+    byte[] bytes = getBytes();
+    int s = 0;
+    int e = bytes.length - 1;
+    // skip all of the space (0x20) in the left side
+    while (s < bytes.length && bytes[s] == 0x20) s++;
+    // skip all of the space (0x20) in the right side
+    while (e >= 0 && bytes[e] == 0x20) e--;
+
+    if (s > e) {
+      // empty string
+      return UTF8String.fromBytes(new byte[0]);
+    } else {
+      return copyUTF8String(bytes, s, e);
+    }
+  }
+
+  public UTF8String trimLeft() {
+    byte[] bytes = getBytes();
+    int s = 0;
+    // skip all of the space (0x20) in the left side
+    while (s < bytes.length && bytes[s] == 0x20) s++;
+    if (s == bytes.length) {
+      // empty string
+      return UTF8String.fromBytes(new byte[0]);
+    } else {
+      return copyUTF8String(bytes, s, bytes.length - 1);
+    }
+  }
+
+  public UTF8String trimRight() {
+    byte[] bytes = getBytes();
+    int e = bytes.length - 1;
+    // skip all of the space (0x20) in the right side
+    while (e >= 0 && bytes[e] == 0x20) e--;
+
+    if (e < 0) {
+      // empty string
+      return UTF8String.fromBytes(new byte[0]);
+    } else {
+      return copyUTF8String(bytes, 0, e);
+    }
+  }
+
+  public UTF8String reverse() {
+    byte[] bytes = getBytes();
+    byte[] result = new byte[bytes.length];
+
+    int i = 0; // position in byte
+    while (i < numBytes) {
+      int len = numBytesForFirstByte(getByte(i));
+      System.arraycopy(bytes, i, result, result.length - i - len, len);
+
+      i += len;
+    }
+
+    return UTF8String.fromBytes(result);
+  }
+
+  public UTF8String repeat(int times) {
+    byte[] bytes = getBytes();
+    int totalLength = numBytes * times;
+
+    if (totalLength < 0) {
+      totalLength = 0;
+    }
+
+    byte[] newBytes = new byte[totalLength];
+    while (times > 0) {
```

Comment:
But `System.arraycopy` only support copying the source data once, we need to make another call if we want to copy it again, doesn't it?
And compare to the `System.arraycopy`, the overhead of while loop probably is trivial.